### ISSUE TITLE
Hourly API for premium subscription

### ISSUE THREAD

**Issue by jbrachey** [apearson/southern-company-api#39]
Hi! I recently upgraded my Southern Company account to premium, and now the hourly API no longer works for me. After investigating more, I see that the normal graph display for hourly on the Southern Company's My Power Usage page (which I believe is what your API pulls data from) no longer displays data for days since I upgraded. The graph that looked like this up until October 3rd (when I upgraded my account):
![before_hourly](https://user-images.githubusercontent.com/65885692/194928589-03d3a1eb-ef31-412d-a6f7-8ce9c01a24a4.png)
now just says "no data available" for every day after October 3rd, which explains why your hourly api can't get the data for those days. Instead, my hourly power usage is displayed in a line graph on energydirect.com that looks like this: 
![after_hourly](https://user-images.githubusercontent.com/65885692/194929553-b0776979-d38e-404c-8e4f-3b9d722a615f.png)
This is a welcome change since this graph has much more precise data down to every 30 minutes (while the old hourly data was always a multiple of 40 kwh for me for some reason, not very precise). However, I was really hoping to use your api or build something similar on top of your api for the more precise hourly data. Do you think this is a realistic goal, and do you have any tips? 

Right now I'm wondering how you went about getting the URLs you needed to get the data from the Southern Company. If I could find the api URL used to fetch the data displayed in my second attached graph, then I imagine I could extract the data much like you guys do in your current methods. I'm also really hoping that the login/verification process for EnergyDirect is the same as it is for the main Southern Company website. Please let me know if you can help in any way. If we can figure this out, I'd be happy to try to add the premium hourly data fetching to this repo (assuming you would want that). Just let me know! Thanks!

**Comment by apearson** [comment]
Hey @jbrachey, 

I'm down with adding the premium hourly data fetching to this repo.  I'll need to do some investigating before being able to add it though.  

A couple questions: 
- What was the process of signing up for the premium hourly data? Do I need to register on energydirect.com?
- Which package did you signed up for that included the hourly data?
- If the hourly data isn't available on my account would you willing to share the requests coming from your account? 

A couple of answers:
- I think adding hourly data is a realistic goal just from what I've seen of Southern Company api design in the past
- I found the URLs by using the browser dev tools and reverse engineering the requests the browser made to the SoCo servers
- The login/verification process seems to be shared across most SoCo properties (would make sense not to duplicate work) 

Thanks,
Andrew Pearson

**Comment by jbrachey** [comment]
Thanks for the quick response! I signed up for the premium account on energydirect.com. The premium plan comes with interval data which is where that detailed hourly graph is. You can sign up for that plan on EnergyDirect's website and then link your Southern Company account, but I'd be happy to just share my requests. I briefly tried to use the dev tools network page to get the URLs and tried them on Postman but had some difficulties. When I made the requests in Postman I seemed to be getting a gif back but it was just blank when I tried to save it to a file and open it. There's a very good chance I was just doing something wrong or using the wrong URL though.

**Comment by jbrachey** [comment]
Thanks for the quick response! I signed up for the premium account on energydirect.com. The premium plan comes with interval data which is where that detailed hourly graph is. You can sign up for that plan on EnergyDirect's website and then link your Southern Company account, but I'd be happy to just share my requests. I briefly tried to use the dev tools network page to get the URLs and tried them on Postman but had some difficulties. When I made the requests in Postman I seemed to be getting a gif back but it was just blank when I tried to save it to a file and open it. There's a very good chance I was just doing something wrong or using the wrong URL though.

Also, I didn't mean to close and reopen this, hit the wrong button lol. 
Thanks again,
Jacob

**Comment by apearson** [comment]
@jbrachey I attempted to sign up for EnergyDirect but my Alabama Power account is not able to use EnergyDirect.  

The best bet going forward is to look over the network requests and see how they're structured

**Comment by jacob-brachey-joulea** [comment]
Hey! I just wanted to give you an update on this stuff. I was able to track the network requests, get the APIs, and get the EnergyDirect hourly data working! However, I ended up doing it in Python since I'm more familiar with Python. However, I'll go ahead and leave the steps here in case you (or someone else) wants to get it working and incorporate it into your npm package at some point. 

As you guessed, the process for EnergyDirect was fairly similar to the process for normal Southern Company. I first hit the "https://webauth.southernco.com/api/login" POST API and use that response to get an ScWebToken. Then I use that ScWebToken in a POST call to "https://www.energydirect.com/ACM/DoWebAuth" which returns two cookies we need, one called edAuth and one called edAuth.SoCoJwt. With those cookies and the user's account number, I make a GET call to "https://www.energydirect.com/Account/GetAccountListing", check if the account is premium and if so, extract the AccountId and the service point number. Finally, with ed_auth, edAuth.SoCoJwt, AccountId, ServicePointId, and the start and end dates, I make the POST request to "https://www.energydirect.com/Reporting/GetReport" which finally gives the actual data. 

Hopefully this helps, and I'm sorry I ended up doing it in Python and couldn't contribute to this repo. If you decide you want to add this functionality and have any more questions about what I did, feel free to contact me and I can try to help. Thanks!

**Comment by jbrachey** [comment]
@apearson Hey! I just wanted to give you an update on this stuff. I was able to track the network requests, get the APIs, and get the EnergyDirect hourly data working! However, I ended up doing it in Python since I'm more familiar with Python. However, I'll go ahead and leave the steps here in case you (or someone else) wants to get it working and incorporate it into your npm package at some point.

As you guessed, the process for EnergyDirect was fairly similar to the process for normal Southern Company. I first hit the "https://webauth.southernco.com/api/login" POST API and use that response to get an ScWebToken. Then I use that ScWebToken in a POST call to "https://www.energydirect.com/ACM/DoWebAuth" which returns two cookies we need, one called edAuth and one called edAuth.SoCoJwt. With those cookies and the user's account number, I make a GET call to "https://www.energydirect.com/Account/GetAccountListing", check if the account is premium and if so, extract the AccountId and the service point number. Finally, with ed_auth, edAuth.SoCoJwt, AccountId, ServicePointId, and the start and end dates, I make the POST request to "https://www.energydirect.com/Reporting/GetReport" which finally gives the actual data.

Hopefully this helps, and I'm sorry I ended up doing it in Python and couldn't contribute to this repo. If you decide you want to add this functionality and have any more questions about what I did, feel free to contact me and I can try to help. Thanks!
